flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java

```diff
@@ -677,6 +734,18 @@ static void serializeStreamStateHandle(StreamStateHandle stateHandle, DataOutput
             RelativeFileStateHandle relativeFileStateHandle = (RelativeFileStateHandle) stateHandle;
             dos.writeUTF(relativeFileStateHandle.getRelativePath());
             dos.writeLong(relativeFileStateHandle.getStateSize());
+        } else if (stateHandle instanceof SegmentFileStateHandle) {
+            if (stateHandle instanceof EmptySegmentFileStateHandle) {
+                dos.writeByte(EMPTY_SEGMENT_FILE_HANDLE);
```

Comment:
IIUC, there is not a chance that `EmptySegmentFileStateHandle` is serialized here?

Reply:
This `serializeStreamStateHandle` is used when serialize `EmptyFileMergingOperatorStreamStateHandle` which holds this as its delegated empty stream state handle.
So it's still needed, right ?

Reply:
Ah.... I missed the `serializeStreamStateHandle(stateHandle.getDelegateStateHandle(), dos);` part